File: angular-app/src/app/services/appointment.service.ts

import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Appointment } from '../models/appointment.model';
import { Pet } from '../models/pet.model';

@Injectable({
  providedIn: 'root'
})
export class AppointmentService {

  private apiUrl = 'https://8080-fadffcfdbddbdfbdfacfcddfbedbebb.premiumproject.examly.io/api/appointments';

  constructor(private http: HttpClient) { }

  /**
   * Fetch all appointments from the server.
   */
  getAllAppointments(): Observable<Appointment[]> {
    return this.http.get<Appointment[]>(this.apiUrl);
  }

  /**
   * Fetch a specific appointment by its ID.
   * @param id - The ID of the appointment to fetch.
   */
  getAppointmentById(id: number): Observable<Appointment> {
    return this.http.get<Appointment>(`${this.apiUrl}/${id}`);
  }

  /**
   * Fetch appointments for a specific user by their user ID.
   * @param userId - The ID of the user whose appointments are to be fetched.
   */
  getAppointmentsByUserId(userId: number): Observable<Appointment[]> {
    return this.http.get<Appointment[]>(`${this.apiUrl}/user/${userId}`);
  }

  /**
   * Add a new appointment.
   * @param appointment - The appointment to add.
   */
  addAppointment(appointment: Appointment): Observable<Appointment> {
    return this.http.post<Appointment>(this.apiUrl, appointment);
  }

  /**
   * Update an existing appointment.
   * @param id - The ID of the appointment to update.
   * @param appointment - The updated appointment data.
   */
  updateAppointment(id: number, appointment: Appointment): Observable<Appointment> {
    return this.http.put<Appointment>(`${this.apiUrl}/${id}`, appointment);
  }

  /**
   * Fetch pet details by appointment ID.
   * @param id - The ID of the appointment.
   */
  getPetByAppointmentId(id: number): Observable<Pet> {
    return this.http.get<Pet>(`${this.apiUrl}/${id}/pet`);
  }

  /**
   * Update the status of an appointment.
   * @param appointmentId - The ID of the appointment.
   * @param status - The new status.
   */
  updateAppointmentStatus(appointmentId: number, status: string): Observable<void> {
    console.log(status);
    return this.http.put<void>(`${this.apiUrl}/${appointmentId}/status`,  status );
  }

  /**
   * Delete an appointment.
   * @param id - The ID of the appointment to delete.
   */
  deleteAppointment(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}

File: springboot-app/src/main/java/com/examly/springapp/service/AppointmentService.java
package com.examly.springapp.service;

import java.util.List;
import java.util.Optional;

import com.examly.springapp.exception.AppointmentExceedException;
import com.examly.springapp.exception.AppointmentNotFoundException;
import com.examly.springapp.model.Appointment;
import com.examly.springapp.model.Pet;

/**
 * Service interface for handling appointment-related operations.
 */
public interface AppointmentService {

    /**
     * Retrieves all appointments.
     * 
     * @return A list of all appointments
     */
    List<Appointment> getAllAppointments();

    /**
     * Retrieves appointments by user ID.
     * 
     * @param userId The ID of the user whose appointments are to be retrieved
     * @return A list of appointments belonging to the specified user
     */
    List<Appointment> getAppointmentsByUserId(int userId);

    /**
     * Adds a new appointment.
     * 
     * @param appointment The appointment to be added
     * @return The added appointment
     * @throws AppointmentExceedException
     * @throws AppointmentNotFoundException
     */
    Appointment addAppointment(Appointment appointment) throws AppointmentExceedException, AppointmentNotFoundException;

    /**
     * Retrieves an appointment by its ID.
     * 
     * @param appointmentId The ID of the appointment to be retrieved
     * @return The retrieved appointment, or null if not found
     */
    Appointment getAppointmentById(int appointmentId) throws AppointmentNotFoundException;

    /**
     * Updates an appointment by its ID.
     * 
     * @param appointmentId The ID of the appointment to be updated
     * @param appointment   The updated appointment details
     * @return The updated appointment, or null if not found
     * @throws AppointmentNotFoundException
     */
    Appointment updateAppointmentById(int appointmentId, Appointment appointment) throws AppointmentNotFoundException;

    /**
     * Deletes an appointment by its ID.
     * 
     * @param appointmentId The ID of the appointment to be deleted
     * @return true if the appointment was successfully deleted, false otherwise
     * @throws AppointmentNotFoundException
     */
    boolean deleteAppointmentById(int appointmentId) throws AppointmentNotFoundException;

    public void updateAppointmentStatus(int appointmentId, String status);

    public List<Appointment> getAllAppointmentsWithPet();

    public Optional<Pet> getPetByAppointmentId(int appointmentId);

}